### ISSUE TITLE
Invalid decrypt_error  response connected to unreliability of the Handshake protocol

### ISSUE THREAD

**Issue by pfg666** [eclipse/tinydtls#79]
## Non-conformance Bug affecting usability of the library
**OS**  
Linux, Ubuntu 20
develop branch, commit c20c6323
Affects both clients and servers

**Non-conformance**

According to the RFC, decrypt_error alerts should be generated when:
> A handshake cryptographic operation failed, including being unable
> to correctly verify a signature or validate a Finished message.

None of these happen in the below capture.  Therein, the client  receives an unencrypted HelloVerifyRequest with epoch 1 is received. The message is unencrypted. The fact that it has epoch 1 should mean that it should not be processed/should instead be discarded/ rejected. 

![etinydtls_invalid_decrypt_error_alert](https://user-images.githubusercontent.com/2325013/119168425-29c06200-ba61-11eb-946e-72bad23c6dae.png)

The issue also affects servers. In the capture below, an ClientKeyExchange with epoch 1 elicits a decrypt_error response from the server for no good reason.

![etinydtls_server](https://user-images.githubusercontent.com/2325013/119181877-7eb8a400-ba72-11eb-8d8d-7b63ad31b024.png)

Abruptly terminating the handshake on receiving a record with an epoch for which a session has not been established has a more severe implication. In case Finished (epoch 1) is received before ChangeCipherSpec (epoch 0) or ClientKeyExchange (epoch 0), the TinyDTLS client/server will abruptly terminate the handshake. This can happen in real world due to reordering. A capture of this is shown below exposing the problem on clients. Notice that the ChangeCipherSpec of the server is sent before Finished, to mimic this scenario. Finished appears as an "Encrypted Handshake Message" in the capture, as it cannot be decrypted by Wireshark.

![etinydtls_decrypt_reorder](https://user-images.githubusercontent.com/2325013/119169835-c33c4380-ba62-11eb-8419-4203b98dfd14.png)

**Steps to Reproduce**
I attached files necessary for reproduction using [DTLS-Fuzzer](https://github.com/assist-project/dtls-fuzzer/), a Java-based tool for testing DTLS libraries. Also included in the archive the captures of the interaction show above. DTLS-Fuzzer requires  the JDK for Java 8. On Ubuntu, this can be installed  by running:
`sudo apt-get install openjdk-8-jdk`

Unpack the archive, `cd` to resulting folder and run `bash reproduce.sh test_sequence_decrypt` or `bash reproduce.sh test_sequence_reorder`, while running an instance of Wireshark on the side. The reproduction script will: 
* setup eclipse TinyDTLS and DTLS-Fuzzer
* launch the TinyDTLS client using the 'tests/dtls-client utility;
* launch DTLS-Fuzzer to execute the input sequences 'test_sequence_decrypt' or 'test_sequence_reorder'  which expose the two problems.

I also added the test sequence 'test_sequence_handshake'. This is similar to the reordering sequence, but without the reordering. By executing this sequence via `bash reproduce.sh test_sequence_handshake`, you should complete the handshake with the client.

Note that the problems affect both clients and servers. On request, I can provide reproduction scripts also for servers. I am guessing fixing the problem for clients will also do it for servers.

Thanks!

[reproduction.tar.gz](https://github.com/eclipse/tinydtls/files/6524413/reproduction.tar.gz)

**Comment by pfg666** [comment]
Hi, I can confirm that PR #85 fixes the incorrect decrypt_error issue. I think I know why DTLS-Fuzzer is not able to decrypt. It relies on a version of TLS-Attacker which does not handle raw keys properly. The setup process normally patches TLS-Attacker before installing it, but that is not the case if an unpatched TLS-Attacker was used to reproduce other bugs.  Try:
`rm -r /tmp/*` - this removes cached files used by the reproduction script
`rm -r ~/.m2/repository/de/rub/` - this removes TLS-Attacker from the local mvn store

Remove the 'dtls-fuzzer' directory and run the reproduction script again. Hopefully, this will update TLS-Attacker and now you will complete the handshake.

**Comment by pfg666** [comment]
BTW, PR #85 does not address the reliability issues that arise if Finished arrives before CCS. When running `bash reproduce.sh test_sequence_reorder` I get an unexpected_message alert and the handshake breaks. I think the easiest way to solve this is by, at least in states where a finished is possible, ignoring handshake records whose epoch numbers are 1 higher than the expected.

**Comment by mrdeep1** [comment]
> I think I know why DTLS-Fuzzer is not able to decrypt

I completely cleaned out /tmp and ~/.m2, unpacked the reproduction.tar.gz from this issue into /tmp/test and ran everything again in /tmp/test (changing `sleep 3` to `sleep 10`) using `bash reproduce.sh test_sequence_handshake` and still get a failure.
````
$ bash reproduce.sh test_sequence_handshake
13:49:55 [main] INFO : Main - Noticed that the first argument is a file. Processing it as an argument file.
13:49:55 [main] INFO : Main - Processing command state-fuzzer-client
13:49:55 [main] INFO : Main - Debug operation is engaged
13:49:55 [main] INFO : Main - Running test runner
13:49:59 [main] WARN : WorkflowConfigurationFactory - Unsupported key exchange algorithm: NULL, not creating ClientKeyExchange Message
13:50:00 [Thread-1] INFO : TlsSUL - Initializing transport handler (waiting ClientHello)
13:50:00 [main] INFO : TlsSUL - Start 0
Jun 07 13:50:03 ALRT setsockopt IPV6_PKTINFO: Protocol not available
13:50:03 [Thread-1] INFO : TlsSUL - Initialized transport handler
13:50:03 [main] INFO : AbstractInputExecutor - Executing input HELLO_VERIFY_REQUEST
13:50:03 [main] INFO : GenericReceiveAction - Received Messages (TlsContext{'server', listening on port 20000}): CLIENT_HELLO, CLIENT_HELLO,
13:50:03 [main] INFO : AbstractInputExecutor - Executing input ECDH_SERVER_HELLO
13:50:04 [main] INFO : GenericReceiveAction - Received Messages (TlsContext{'server', listening on port 20000}):
13:50:04 [main] INFO : AbstractInputExecutor - Executing input CERTIFICATE
13:50:04 [main] INFO : GenericReceiveAction - Received Messages (TlsContext{'server', listening on port 20000}):
13:50:04 [main] INFO : AbstractInputExecutor - Executing input ECDH_SERVER_KEY_EXCHANGE
13:50:04 [main] INFO : GenericReceiveAction - Received Messages (TlsContext{'server', listening on port 20000}):
13:50:04 [main] INFO : AbstractInputExecutor - Executing input SERVER_HELLO_DONE
13:50:04 [main] INFO : GenericReceiveAction - Received Messages (TlsContext{'server', listening on port 20000}):
13:50:04 [main] INFO : AbstractInputExecutor - Executing input CHANGE_CIPHER_SPEC
13:50:05 [main] INFO : GenericReceiveAction - Received Messages (TlsContext{'server', listening on port 20000}): ECDH_CLIENT_KEY_EXCHANGE, CHANGE_CIPHER_SPEC, FINISHED,
13:50:05 [main] INFO : AbstractInputExecutor - Executing input FINISHED
Jun 07 13:50:05 WARN decryption failed
13:50:05 [main] INFO : GenericReceiveAction - Received Messages (TlsContext{'server', listening on port 20000}): Alert(FATAL,DECRYPT_ERROR),
13:50:05 [main] INFO : AbstractInputExecutor - Executing input APPLICATION
Jun 07 13:50:05 WARN no peer available, send an alert
13:50:05 [main] INFO : GenericReceiveAction - Received Messages (TlsContext{'server', listening on port 20000}):
13:50:05 [main] INFO : TestRunner - Test: HELLO_VERIFY_REQUEST ECDH_SERVER_HELLO CERTIFICATE ECDH_SERVER_KEY_EXCHANGE SERVER_HELLO_DONE CHANGE_CIPHER_SPEC FINISHED APPLICATION
1 times outputs: CLIENT_HELLO+ TIMEOUT TIMEOUT TIMEOUT TIMEOUT ECDH_CLIENT_KEY_EXCHANGE|CHANGE_CIPHER_SPEC|FINISHED Alert(FATAL,DECRYPT_ERROR) TIMEOUT

Terminated
````
And the dtls-client complains with (in area of first issue) (with #86 for better logging)
````
Jun 07 14:40:33 DEBG peer: 127.0.0.1:20000
Jun 07 14:40:33 DEBG bytes from peer: (14 bytes): 14FEFD0000000000000005000101
Jun 07 14:40:33 DEBG dtls_handle_message: FOUND PEER
Jun 07 14:40:33 DEBG got 'change_cipher_spec' epoch 0 sequence 5 (14 bytes)
Jun 07 14:40:33 DEBG new bitfield is ffffffffffffffff sequence base 5
Jun 07 14:40:33 DEBG receive header: (13 bytes):
00000000 14 FE FD 00 00 00 00 00  00 00 05 00 01
Jun 07 14:40:33 DEBG receive unencrypted: (1 bytes):
00000000 01
Jun 07 14:40:34 DEBG peer: 127.0.0.1:20000
Jun 07 14:40:34 DEBG bytes from peer: (37 bytes): 16FEFD000100000000000000181400000C000500000000000C064DD5ED5DDA5831713B906E
Jun 07 14:40:34 DEBG dtls_handle_message: FOUND PEER
Jun 07 14:40:34 DEBG got 'handshake' epoch 1 sequence 0 (37 bytes)
Jun 07 14:40:34 DEBG nonce: (16 bytes): 5E715A311400000C0005000000000000
Jun 07 14:40:34 DEBG key: (16 bytes): 95C6B1A67BD4EF8330E6ED68F655E433
Jun 07 14:40:34 DEBG ciphertext: (16 bytes): 0000000C064DD5ED5DDA5831713B906E
Jun 07 14:40:34 WARN decryption failed
Jun 07 14:40:34 INFO decrypt_verify() failed
Jun 07 14:40:34 DEBG dtls_prepare_record(): encrypt using TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8
Jun 07 14:40:34 DEBG nonce:: (16 bytes): 51EFB440000100000000000100000000
Jun 07 14:40:34 DEBG key:: (16 bytes): E53038E73F07FA0238951028F8973392
Jun 07 14:40:34 DEBG message:: (18 bytes): 00010000000000011A628E79239137EF67D9
Jun 07 14:40:34 DEBG send header: (13 bytes):
00000000 15 FE FD 00 01 00 00 00  00 00 01 00 12
Jun 07 14:40:34 DEBG send unencrypted: (2 bytes):
00000000 02 33
````

**Comment by pfg666** [comment]
Could you replace 'dtls-fuzzer/target/dtls-fuzzer.jar' by the  [jar](https://github.com/pfg666/reproduction/blob/main/dtls-fuzzer.jar) I have just uploaded and re-run the reproduction script?

**Comment by mrdeep1** [comment]
Unfortunately my dtls-client still fails to decrypt on receipt of the finished message from the server.

**Comment by pfg666** [comment]
Ok, let's try one more thing. Unpack/build the PR [files](https://github.com/eclipse/tinydtls/archive/5d165d31dac23759b9034c608c89d0d8636a83dc.tar.gz) into 'etinydtls', replacing existing contents, then re-run. If this does not work, then I don't know what will. I can reproduce the issue both on my main host and on a VM I set up for this purpose.

**Comment by mrdeep1** [comment]
Again, unfortunately no change.  I will try and dig deeper at my end.

$ md5sum dtls-fuzzer/target/dtls-fuzzer.jar
8ebc1dc935e2c5d1fe35403b17bda954  dtls-fuzzer/target/dtls-fuzzer.jar

**Comment by pfg666** [comment]
I think it is likely that something is off with raw key handling in DTLS-Fuzzer. Support for that is very shaky. Hence, I am providing reproduction using PSK. See if this works, the process is the same. Note that I am using a 'dtls-client.c' that is modified such that it echoes data and uses a specific psk (1234).
[reproduction.tar.gz](https://github.com/eclipse/tinydtls/files/6610751/reproduction.tar.gz)

In the reproduction script, I am downloading the PR. All the test files have been updated for PSK.

**Comment by mrdeep1** [comment]
PSK works as expected.  Now I have something to be able to debug!

My suspicion is that the raw key of 32 bytes when represented as an ASN.1 value being passed over the DTLS channel (which has to be a positive integer) ,ignoring the integer type and length part, can be 33 bytes long as a leading 0x00 need to be inserted if the first of the 32 bytes is >= 0x80 to keep it positive (see #58).

**Comment by boaks** [comment]
AFAIK, the encoded ec public key is not build on asn.1 integers.

e.g.:
```
3059301306072A8648CE3D020106082A8648CE3D03010703420004987B0020640441C4AFC88B4D777CCB59ABFEFA38C46970F17D5214652F6111F1A3164591846E187156DBBABD2813225F2F0DCE657E6A27E288A0B77D01BB50AE
```

The BITSTRING starts with `03` (tag) followed by the length (one byte, `42`), followed by the number of not used bits (one byte, `00`), followed by `04` (uncompressed) and fix 64 bytes for the ec public key (for secp256r1).

The encoding of a signature is different.

**Comment by boaks** [comment]
I'm not sure, what exactly is tested. A basic RPK handshake without client authentication seems to be possible with a tinydtls-client and a Californium-server.

[rpk.pcapng.gz](https://github.com/eclipse/tinydtls/files/6612061/rpk.pcapng.gz)

If you describe, what the difference should be, I may check that difference as well.
(Or should a specific PR be considered?)

@pfg666 

Did you try to check that example also against a Californium client?

**Comment by mrdeep1** [comment]
Yes, I meant the ASN.1 signature.

The test is dtls-client against DTLS-fuzzer acting as a server, following the instructions in the initial comment.  What fails for me is  `reproduce.sh test_sequence_handshake` which should be working (I fail in handling the server 'finished'.  The updated `dtls-fuzzer/target/dtls-fuzzer.jar` made no difference.

**Comment by pfg666** [comment]
@boaks  When testing Scandium, we use x509 Certificates which are better supported. 

To check where the problem lies (TinyDTLS or DTLS-Fuzzer), I suggest we compare two de-randomized captures, one on my machine, and the other on @mrdeep1's. De-randomization in DTLS-Fuzzer is done by setting `useFreshRandom` to false in the 'sul.config' file (e.g. ` <useFreshRandom>false</useFreshRandom>` ). In TinyDTLS de-randomization can be achieved by altering the `dtls_prng` function. 

To make things easy, I attached reproduction files plus the capture I get when running `reproduce.sh test_sequence_handshake` twice. I am running this on Linux, so in my case, I de-randomize by changing 'dtls_prng_posix.c' by filling the random buffer with 1.  I also simplified the reproduction script, making it download the DTLS-Fuzzer .jar as opposed to building the tool. I should have  done this from the offset to avoid maven/build issues.

Run `reproduce.sh test_sequence_handshake` twice (to ensure de-randomization takes place), compare with the capture 'capture_derand.pcap' and check where the differences lie (is it in what the client or the fuzzer generates). 

[derand_reproduction.tar.gz](https://github.com/eclipse/tinydtls/files/6615147/derand_reproduction.tar.gz)

**Comment by boaks** [comment]
> Yes, I meant the ASN.1 signature.

Wouldn't the test then need to cherry-pick PR #58 ?

> following the instructions in the initial comment. 

I'm not sure: is that the case, where the server exchanged the order CCS and the FINISH?

**Comment by pfg666** [comment]
@boaks This is the normal case of simply completing a handshake using ECDSA raw keys. I can complete it on both my main machine, and on a test VM, but not @mrdeep1 . In any case, we were able to reproduce the reliability issue (FINISH received before CCS breaking handshake) using a PSK handshake. Still curious to know whether the latest reproduction files elucidate the problem.

**Comment by mrdeep1** [comment]
> To check where the problem lies (TinyDTLS or DTLS-Fuzzer)

Will do this on my Linux Ubuntu 20 system

> Wouldn't the test then need to cherry-pick PR #58 ?

Fair comment - will try that at some point.  It will be good when that is merged.

> I'm not sure: is that the case, where the server exchanged the order CCS and the FINISH?

That is true when using parameter `test_sequence_reorder` - `test_sequence_handshake` has everything in the correct order.

**Comment by boaks** [comment]
> This is the normal case of simply completing a handshake using ECDSA raw keys.

With PR #58, that should work. That was at least the test I did.

**Comment by mrdeep1** [comment]
I have found that if the Finished packet (epoch 1 seq 0) is ignored before receiving the CCS, the following application data is correctly decrypted - BUT, the Finished check never takes place which is bad news.  The Finished packet does need to get buffered somehow, but that is not straight forward.

**Comment by boaks** [comment]
> I have found that if the Finished packet (epoch 1 seq 0) is ignored before receiving the CCS, the following application data is correctly decrypted - BUT, the Finished check never takes place which is bad news. The Finished packet does need to get buffered somehow, but that is not straight forward.

That will be an "follow up" error.

 I'm still confused:
Does the basic handshake (no reorder) work? Or not?

**Comment by mrdeep1** [comment]
Basic PRK handshake no reorder still fails for me - even when #58 is used for the original reproduction.tar.gz.  Building derand_reproduction.tar.gz now.

**Comment by boaks** [comment]
> Basic PRK handshake no reorder still fails for me 

Fails with the TLS-Attacker? And the error is the decrypt error of the FINISH?
It succeeds with Californium, so I'm not sure, if that is really a bug of tinydtls.
Maybe, checking it tinydtls with guntls makes sense.

**Comment by pfg666** [comment]
@boaks We use x509 certificates with Scandium. With TinyDTLS we use raw keys, for which support is faulty in the version of TLS-Attacker DTLS-Fuzzer is using. I think it is very conceivable that DTLS-Fuzzer/the reproduction scripts/differing environments are to blame. But I think the issue should be clarified once we know what @mrdeep1  gets after running the de-randomized reproduction script.

**Comment by mrdeep1** [comment]
With de-randomized, it works or fails when running  `bash reproduce.sh test_sequence_handshake` multiple times. 
[capture.zip](https://github.com/eclipse/tinydtls/files/6615699/capture.zip) holds a broken capture.

**Comment by boaks** [comment]
Do you have also a "good one" :-). Please upload.

**Comment by boaks** [comment]
That's life ...

**Comment by boaks** [comment]
A very first guess: 
The attacker sends the CCS too early (before the the clients response flight), ... that may mixes up a lot.
But I don't understand (for now), why that affects the FINISH's MAC.

**Comment by mrdeep1** [comment]
[capture_bad_good.zip](https://github.com/eclipse/tinydtls/files/6615836/capture_bad_good.zip)  Ignore first 5 records - I rebuilt all from scratch but forgot to update the sleep time before starting dtls-client in reproduce.sh.

**Comment by mrdeep1** [comment]
Sorry - good starts at record 20 which has the CCS where I expect it.  Though I don't see how the hash is messed up (yet).

**Comment by boaks** [comment]
Please copy the log output of tinydtls on failure to see, what exactly is reported. Is it still the same as above?

> Jun 07 14:40:34 DEBG dtls_handle_message: FOUND PEER
> Jun 07 14:40:34 DEBG got 'handshake' epoch 1 sequence 0 (37 bytes)
> Jun 07 14:40:34 DEBG nonce: (16 bytes): 5E715A311400000C0005000000000000
> Jun 07 14:40:34 DEBG key: (16 bytes): 95C6B1A67BD4EF8330E6ED68F655E433
> Jun 07 14:40:34 DEBG ciphertext: (16 bytes): 0000000C064DD5ED5DDA5831713B906E
> Jun 07 14:40:34 WARN decryption failed

**Comment by pfg666** [comment]
Looking at the capture, I think I know where the problem is. The problem is in the timing parameters. DTLS-Fuzzer in the first attempted handshake after the 5 records, sends ChangeCipherSpec before receiving ClientKeyExchange from the client. Whenever it sends ChangeCipherSpec, it generates keys from existing material. Of course, since it hasn't received the ClientKeyExchange, the generated keys will be bogus. 

In 'dtls-fuzzer.args' increase the '-timeout' value (which is now set 100), to something bigger (say 300), and I think now you should be able to reproduce also with ECDSA. Next time I will use much more conservative timing parameters for the purpose of reproduction.

**Comment by boaks** [comment]
Maybe the TLS - Attacker sends the `FINISH` sometimes not encrypted?

If it fails, the record has 37 bytes => 24 byte handshake message

```
Record header: 16FEFD00010000000000000018

handshake message: 1400000C000500000000000C064DD5ED5DDA5831713B906E
```

**Comment by boaks** [comment]
But the cause, why tinydtls fails ...

**Comment by mrdeep1** [comment]
Changing timeout to 300 makes it work every time for `test_sequence_handshake`.  I am aware that computing the hash for each record is slow and compute intensive, compared to GnuTLS where the whole exchanged records are put into a concatenated buffer and then the hash done once across all the records.  But that fix is for another day.

Doing `test_sequence_reorder `fails as expected, but using #85 code it now works for me.

**Comment by pfg666** [comment]
@boaks The only option in DTLS-Fuzzer currently available is to adjust the response timeout parameter (the amount of time the fuzzer waits for a response to a sent record, before sending the next record) according to the hardware on which the fuzzer/sut is running and of course, according to any network-induced delays (which are 0 in our case, since we test over loopback). 

The response timeout value we use for TinyDTLS is 100 ms, which for us has always been more than enough, but appears not to be enough for @mrdeep1 .  We normally use a rather low timeout value so as to maximize the number of tests we can execute in a given time budget.

**Comment by boaks** [comment]
@pfg666 

For me this seems to be obvious, that a tool, which wants to test "exchanges of messages" must have the feature to wait for messages of the other. We will see, how much headache this will cause.

**Comment by pfg666** [comment]
@mrdeep1 The #85 code results in segmentation fault for me when running `test_sequence_handshake`. 

@boaks DTLS-Fuzzer is not really designed for executing a suite of tests, but for checking the behavior to arbitrary sequences of inputs. In such a context, you never wait for specific messages (since you don't know what you are supposed to receive), but instead wait a fixed amount of time.  TLS-Attacker, on the other hand, does provide means of defining/executing test suites. I checked and indeed, it is possible in a TLS-Attacker workflow (aka test case for our purpose) to have TLS-Attacker wait for a certain message, before sending subsequent records. That is useful to know for future reproduction cases. 

It seems progress can be made on debunking the problem further without intervention on my part. However, if the need arises, I can provide reproduction where expected messages are waited for.

**Comment by mrdeep1** [comment]
Apologies # 85 has been updated to fix segmentation fault - sorry about that.

I am hosting DTLS-Fuzzer using Ubuntu 20 with 2 CPUs on a ESX5.5 server which has 2.3GHz CPUs - somewhat old but does what I need it to do.  Some of my testing was done across a local network as my github source was on another VM.

Using valgrind for dtls-client is a non starter - even if timeout is set to 1000.

Not sure that expected messages would necessarily help here in getting to the bottom of things (which I think we have done now) (apart from buffering the Finished until CCS has arrived).

**Comment by boaks** [comment]
@pfg666 

Is it possible to retransmit a flight?

> HELLO_VERIFY_REQUEST
> ECDH_SERVER_HELLO
> CERTIFICATE
> ECDH_SERVER_KEY_EXCHANGE
> SERVER_HELLO_DONE
> @CHANGE_CIPHER_SPEC[{@@type:RECORD_DEFER}]
> @FINISHED[{@@type:RECORD_FLUSH}]
> APPLICATION

Assuming, a implementation drops the FINISH before CCS, RFC6347 considers the server to resend the flight 6 with new record numbers. Is that possible?

**Comment by boaks** [comment]
@pfg666 

Just in the case, you're able to spend more time in it:
`develop` contains now a lot of fixes. Maybe the time to retest.
Otherwise, could this issue be closed?

**Comment by pfg666** [comment]
@boaks I can confirm that the decrypt_error alert is not sent anymore. Regarding robustness to record reordering, my testing shows mixed results. On the one hand, TinyDTLS does not seem to drop the connection anymore on receiving FINISHED before CCS. However, there are some problems. I will switch to the TinyDTLS server, since the problem is easier to explain there.

Here is the interaction I am getting when testing the TinyDTLS server via the dtls-server utility, using the PSK ciphersuite TinyDTLS supports.

->CLIENT_HELLO
<-HELLO_VERIFY_REQUEST
->CLIENT_HELLO
<-SERVER_HELLO, SERVER_HELLO_DONE
->PSK_CLIENT_KEY_EXCHANGE, FINISHED, CHANGE_CIPHER_SPEC (out-of-order), CHANGE_CIPHER_SPEC, FINISHED 
(nothing from the server)
->APPLICATION
(the server decrypts and displays the decrypted application data, however it does not echo it)

The problem is that the server does generate CHANGE_CIPHER_SPEC, FINISHED after receiving the retransmission. Not only that, but it processes the application data in a sense (it decrypts it), despite not truly completing the handshake. For clients, it is a bit more difficult to expose this problem, but initial testing shows it is still there.

Here is what I suggest we do, perhaps I can close this issue (since the Alert is no longer sent), and post another one strictly on the handling of reordered messages (where I give reproduction scripts and all)? The issue was a bit too broad to begin with perhaps. I will need to look at the problem again to ensure it is not DTLS-Fuzzer, though so far, the other SUTs I have tested seem to handle the situation of reorderd packets well.

**Comment by boaks** [comment]
Would it be possible to get a capture?

**Comment by boaks** [comment]
I guess, no new record_sequence_numbers are used (the capture would verify that).
Therefore only the CCS is processed, not the second FINISH.
On a first look, I don't see, what prevents tinydtls from processing application data without FINISH, but that would be for sure a bug. Maybe I oversee something.
So, if possible, please a capture.

**Comment by pfg666** [comment]
I attached a .zip, which contains captures of two handshakes with a TinyDTLS server wherein the last flight from the client is received out-of-order, and is re-transmitted, with the re-transmission arriving in order. I also attached a capture generated by executing the same flows on MbedTLS. If I am not mistaken, Scandium also handles this situation fine. Besides that, I also included a regular handshake with the TinyDTLS server. By default, arbitrary Application data should be echoed after the handshake is completed. What TinyDTLS does in the out-of-order case is that it just prints it out.

[out-of-order_handshakes.zip](https://github.com/eclipse/tinydtls/files/7758680/out-of-order_handshakes.zip)

The PSK is "1234" (to use to decrypt in Wireshark).

**Comment by boaks** [comment]
Thanks.

The record seqn are OK, so my first guess doesn't apply.
FMPOV, I would move it to a new issue.

(Is the PSK "1234" ascii, so in hex "31323334"?)

**Comment by boaks** [comment]
```
      dtls_debug("decrypt_verify(): found %i bytes cleartext\n", clen);
      dtls_security_params_free_other(peer);
      dtls_debug_dump("cleartext", *cleartext, clen);
```

Maybe, if possible, add the output of processing the FINISH message as well.

**Comment by boaks** [comment]
Using Californium to reproduce the failure in processing unordered ccs/finish, my result is a successful handshake.

[my_tinydtls_unordered.pcapng.gz](https://github.com/eclipse/tinydtls/files/7760605/my_tinydtls_unordered.pcapng.gz)

Maybe the Finish in you test is wrong? If possible please check the tinydtls server logs as mentioned in the comment before this.

**Comment by boaks** [comment]
unordered - ordered, also success. 

[my_tinydtls_unordered2.pcapng.gz](https://github.com/eclipse/tinydtls/files/7760678/my_tinydtls_unordered2.pcapng.gz)

**Comment by boaks** [comment]
It's really irritating, that your test passes for mbedtls, but not for tinydtls.

Are your "public tools" "up to date", so that I have a chance (with reasonable time), to reproduce the effects and get my own capture and logs?

**Comment by pfg666** [comment]
Hi, I have been looking into this and frankly it has me puzzled. I am never surprised to find bugs in DTLS-Fuzzer, but here I am unsure. I de-randomized TinyDTLS (by making the cookie and prg return 0's), and did the same to DTLS-Fuzzer (using TLS-Attacker's built-in option). This should ensure the keys are the same, and that the Finished required from the client over the two (regular and with out-of-order flight) handshakes is exactly the same. 

[derandomized.pcapng.gz](https://github.com/eclipse/tinydtls/files/7763502/derandomized.pcapng.gz)

If you look at the capture, DTLS-Fuzzer produces the the same Finished messages in both scenarios. Yet in the regular case the handshake is completed, while in the out-of-order case it is not. TinyDTLS is unable to decrypt the Finished message. I will try to capture the application logs. It may be timing specific whether the handshake succeeds or not.

**Comment by boaks** [comment]
if you run tinydtls with logging enabled (I only use it via libcoap, there it's `-v 7`, I get some output on the console. Is it possible, that you provide that?

**Comment by boaks** [comment]
Which tinydtls commit are you using?

**Comment by boaks** [comment]
In my test, the console shows

> Dec 22 16:40:15.899 DEBG *  [::ffff:127.0.0.1]:5684 <-> [::ffff:127.0.0.1]:40169 (if1) DTLS: received 53 bytes
> Dec 22 16:40:15 INFO received message of type 22, epoch 1
> Dec 22 16:40:15 DEBG dtls_handle_message: FOUND PEER
> Dec 22 16:40:15 DEBG got 'handshake' epoch 1 sequence 0 (53 bytes)
> Dec 22 16:40:15 WARN No security context for epoch: 1
> Dec 22 16:40:15 INFO decrypt_verify() failed, drop message.
> Dec 22 16:40:15.900 DEBG *  [::ffff:127.0.0.1]:5684 <-> [::ffff:127.0.0.1]:40169 (if1) DTLS: received 14 bytes
> Dec 22 16:40:15 INFO received message of type 20, epoch 0
> Dec 22 16:40:15 DEBG dtls_handle_message: FOUND PEER
> Dec 22 16:40:15 DEBG got 'change_cipher_spec' epoch 0 sequence 3 (14 bytes)
> Dec 22 16:40:15 DEBG new bitfield is ffffffffffffffff sequence base 3
> Dec 22 16:40:15 DEBG receive header: (13 bytes):
> 00000000 14 FE FD 00 00 00 00 00  00 00 03 00 01 
> Dec 22 16:40:15 DEBG receive unencrypted: (1 bytes):
> 00000000 01 
> Dec 22 16:40:15.900 DEBG got psk_identity: 'Client_identity'
> Dec 22 16:40:15 DEBG psk: (9 bytes):
> 00000000 73 65 63 72 65 74 50 53  4B 
> Dec 22 16:40:15 DEBG client_random: (32 bytes): 61C346DF51B7C942EA30E0032830CC25BA6B1486C8139996B9E2F94E7E15BF47
> Dec 22 16:40:15 DEBG server_random: (32 bytes): 00000000066D500D8F1569050CE6861D8D28F1A992213098B7A9C4E7CDEC73D8
> Dec 22 16:40:15 DEBG pre_master_secret: (22 bytes): 0009000000000000000000000973656372657450534B
> Dec 22 16:40:15 DEBG extended_master_secret: (48 bytes): C5070735AA3C5BBD35E12477C370C761F759865163111E8243A16837466DE64E72F524A549F1ACFEB8D2182BA098C5BC
> Dec 22 16:40:15 DEBG key_block (40 bytes):
> Dec 22 16:40:15 DEBG   client_MAC_secret: (0 bytes): 
> Dec 22 16:40:15 DEBG   server_MAC_secret: (0 bytes): 
> Dec 22 16:40:15 DEBG   client_write_key: (16 bytes): 7DCBC73883CE97D558E2FA766A6C4CE8
> Dec 22 16:40:15 DEBG   server_write_key: (16 bytes): 1226829A44D56B93237654848DA47D1D
> Dec 22 16:40:15 DEBG   client_IV: (4 bytes): FFEE7CAB
> Dec 22 16:40:15 DEBG   server_IV: (4 bytes): EA1351E5
> Dec 22 16:40:17.901 DEBG *  [::ffff:127.0.0.1]:5684 <-> [::ffff:127.0.0.1]:40169 (if1) DTLS: received 14 bytes
> Dec 22 16:40:17 INFO received message of type 20, epoch 0
> Dec 22 16:40:17 DEBG dtls_handle_message: FOUND PEER
> Dec 22 16:40:17 DEBG got 'change_cipher_spec' epoch 0 sequence 5 (14 bytes)
> Dec 22 16:40:17 ALRT No security context for epoch: 0
> Dec 22 16:40:17 INFO decrypt_verify() failed, drop message.
> Dec 22 16:40:17.901 DEBG *  [::ffff:127.0.0.1]:5684 <-> [::ffff:127.0.0.1]:40169 (if1) DTLS: received 53 bytes
> Dec 22 16:40:17 INFO received message of type 22, epoch 1
> Dec 22 16:40:17 DEBG dtls_handle_message: FOUND PEER
> Dec 22 16:40:17 DEBG got 'handshake' epoch 1 sequence 1 (53 bytes)
> Dec 22 16:40:17 DEBG nonce: (16 bytes): FFEE7CAB000100000000000100000000
> Dec 22 16:40:17 DEBG key: (16 bytes): 7DCBC73883CE97D558E2FA766A6C4CE8
> Dec 22 16:40:17 DEBG ciphertext: (32 bytes): D540616366A5C1ECA3D44632071BFAD8AD3B705B43D58175210F06C0DB1D8179
> Dec 22 16:40:17 DEBG decrypt_verify(): found 24 bytes cleartext
> Dec 22 16:40:17 DEBG cleartext: (24 bytes): 1400000C000300000000000CDFB11AF204F4C123001E21C7
> Dec 22 16:40:17 DEBG new bitfield is 1 sequence base 1
> Dec 22 16:40:17 DEBG receive header: (13 bytes):
> 00000000 16 FE FD 00 01 00 00 00  00 00 01 00 28 
> Dec 22 16:40:17 DEBG receive unencrypted: (24 bytes):
> 00000000 14 00 00 0C 00 03 00 00  00 00 00 0C DF B1 1A F2 
> 00000010 04 F4 C1 23 00 1E 21 C7  
> Dec 22 16:40:17 DEBG received handshake packet of type: finished (20)
> Dec 22 16:40:17 DEBG handle handshake packet of type: finished (20)
> Dec 22 16:40:17 DEBG d:: (12 bytes): DFB11AF204F4C123001E21C7
> Dec 22 16:40:17 DEBG v:: (12 bytes): DFB11AF204F4C123001E21C7
> Dec 22 16:40:17 DEBG add MAC data: (24 bytes): 1400000C000300000000000CDFB11AF204F4C123001E21C7

The first Finish is dropped for wrong epoch.
The first CCS is processed.
The second CCS is dropped for wrong epoch.
The second Finish is processed.

What does your console show?

**Comment by pfg666** [comment]
See attachment. It contains both the capture and the logs.
[capture_and_logs.tar.gz](https://github.com/eclipse/tinydtls/files/7763726/capture_and_logs.tar.gz)

Just to ensure we are testing the right thing, the commit number is c8d2def.

**Comment by boaks** [comment]
regular.log has 0 bytes

**Comment by boaks** [comment]
The out-of-order.log looks as on my side. Except, that the decrypt doesn't work (on my side it works).

**Comment by pfg666** [comment]
[regular.log](https://github.com/eclipse/tinydtls/files/7763778/regular.log)

**Comment by boaks** [comment]
I switched to the commit you're using.
Lets change some logs in order to ensure, we really test with that.

L3353:
    dtls_alert("No security context for epoch: %i\n", dtls_get_epoch(header));
to:
    dtls_alert("Decrypt: no security context for epoch: %i\n", dtls_get_epoch(header));

L4244:
      dtls_warn("No security context for epoch: %i\n", dtls_get_epoch(header));
to
      dtls_warn("Decode: no security context for epoch: %i\n", dtls_get_epoch(header));

**Comment by boaks** [comment]
I get (unordered):

Dec 22 17:19:15 DEBG got 'handshake' epoch 1 sequence 0 (53 bytes)
Dec 22 17:19:15 WARN Decode: no security context for epoch: 1

Dec 22 17:19:17 DEBG got 'change_cipher_spec' epoch 0 sequence 5 (14 bytes)
Dec 22 17:19:17 ALRT Decrypt: no security context for epoch: 0

**Comment by pfg666** [comment]
Come to think of it, I am sure the problem is on my end. The premise that Finished would stay identical in a retransmission is false. If I am not mistaken, encryption/authentication includes the record sequence number. And of course, if the record sequence number is increased, then the ciphertext will necessarily change.

**Comment by pfg666** [comment]
So TinyDTLS is right to invalidate the second Finished, since it uses a different sequence number to the first, yet it has the same cipher text. The reason I could run the test successfully on other implementations is because other implementations were caching the first Finished (which was correctly generated). It may still perhaps be a problem that TinyDTLS prints out the derypted Application text, but that is orthogonal to this issue, which based on your unit tests, should be fixed.

So if you agree, I think we can close the issue.

**Comment by boaks** [comment]
Not sure. Though you keep the nonce (instead of insert the epoch/rseqn), that should not really count. I ahev to think more about that.

**Comment by pfg666** [comment]
I am sure the cipher text should change based on the record sequence number, otherwise the implementation would be susceptible to replay attacks. The MAC is computed also over the record sequence number, and it will fail if this is changed, which is exactly what I was doing in my test cases. Therein, I was simply re-sending the old Finished ciphertext and only adjusting the record sequence number, causing the MAC validation step to fail.

**Comment by boaks** [comment]
Yes, sure. Usually for CCM you copy the epoch/rseqn to the nonce. You didn't. 
Decrypt is using that nonce, not the epoch/rseqn.

**Comment by boaks** [comment]
Could you adjust the logging messages in order to verify, that you test that build?

**Comment by pfg666** [comment]
Do you mean these:

> L3353:
dtls_alert("No security context for epoch: %i\n", dtls_get_epoch(header));
to:
dtls_alert("Decrypt: no security context for epoch: %i\n", dtls_get_epoch(header));

L4244:
dtls_warn("No security context for epoch: %i\n", dtls_get_epoch(header));
to
dtls_warn("Decode: no security context for epoch: %i\n", dtls_get_epoch(header));

I did, but I am unsure how these are going to ensure that I am testing the right build.

**Comment by boaks** [comment]
Ah, your right! That's the aad, I missed that.

Anyway ...
- issue #114
- I found also two more code smells. I will prepare a PR

**Comment by pfg666** [comment]
Yes, I informed @obgm a long time ago about the processing of Application data problem. But thought that was a bit too sensitive to have an issue posted on it. I think the problem resurfaces in the out-of-order.log here.

**Comment by boaks** [comment]
> I did, but I am unsure how these are going to ensure that I am testing the right build.

The idea was, that the commit is checked out, but actually a different binary is used. So, if you adapt the logging messages, you see either the new one (all ok), or the old one (binary under test is different).

**Comment by pfg666** [comment]
ah, I got you. Yeap, I get the same behavior even when seeing the updated logs. Fortunately, the behavior seems to be the correct behavior. On your call, I can close the issue, since it seems to be fixed. It will give me a bit to think about next time I make test cases.

**Comment by boaks** [comment]
I think, this is done.

If you have time left, maybe you check your old issue #13 as well?

**Comment by pfg666** [comment]
Both decrypt_error response and Handshake protocol (un-)reliability appear to be fixed.
